src/sas/system/config/schema_elements.py

```diff
@@ -0,0 +1,236 @@
+
+
+class SchemaBool(SchemaVariable):
+    schema_variable_type = "bool"
+
+    def coerce(self, value):
+        if isinstance(value, bool):
+            return value
+        else:
+            raise CoercionError(f"Cannot coerce {type(value)} to bool")
+
+
+class SchemaInt(SchemaVariable):
+    schema_variable_type = "int"
+
+    def coerce(self, value):
+        if isinstance(value, int):
+            return value
+        else:
+            raise CoercionError(f"Cannot coerce {type(value)} to int")
+
+
+class SchemaFloat(SchemaVariable):
+    schema_variable_type = "float"
+
+    def coerce(self, value):
+        if isinstance(value, (int, float)):
+            return float(value)
+        else:
+            raise CoercionError(f"Cannot coerce {type(value)} to bool")
+
+
+class SchemaStr(SchemaVariable):
+    schema_variable_type = "str"
+
+    def coerce(self, value):
+        return str(value)
+
+
+
+class SchemaNonSpecified(SchemaElement):
+    """ Representation of a list with elements of an unknown type -
+    we use this when an empty list is in the config, or default is set to None"""
+
+    def __init__(self):
+        pass
+
+    def coerce(self, value):
+        return value
+
+    def __eq__(self, other: SchemaElement):
+        return isinstance(other, SchemaNonSpecified)
+
+    def __repr__(self):
+        return f"SchemaNonSpecified"
+
+
+class SchemaList(SchemaElement):
+    """ Schema Element representing a homogeneous list"""
+    def __init__(self, child_type: SchemaElement):
+        self.child_type = child_type
+
+    def coerce(self, value):
+        # Only really need to check for list as this should be the only json object made, but let's make it more general
+        if not isinstance(value, (tuple, set, list)):
+            raise CoercionError(f"Cannot coerce {value} ({type(value)}) to list")
+
+        if not isinstance(value, list):
+            logger = logging.getLogger(self.__class__.__name__)
+            logger.warning(f"Corercing variable of type {type(value)} to list ({value})")
+
+        return [self.child_type.coerce(x) for x in value]
+
+    def __eq__(self, other: SchemaElement):
+        if isinstance(other, SchemaList):
+            return self.child_type == other.child_type
+        else:
+            return False
+
+    def __repr__(self):
+        return f"SchemaList({repr(self.child_type)})"
+
+
+def create_schema_element(name: str, value, recursion_depth: int=10) -> SchemaElement:
+    """ Get an appropriate schema element for a specified config datum"""
+
+    # Limits the depth of list config items
+    if recursion_depth <= 0:
+        raise SchemaError(f"Config element '{name}' is too nested, or is self-referential")
+
+    if value is None:
+        logger = logging.getLogger("create_schema_element")
+        logger.warning(f"Non-specified type for variable '{name}'")
+        return SchemaNonSpecified()
+
+    # List case
+    elif isinstance(value, list):
```

Comment:
SchemaList coerces tuples and sets to a list (with a warning), but your creation method only accepts lists. Any reason for this?

Reply:
If I understand you correctly, you're asking why when parsing the fields in `Config` it only accepts lists, but one can coerce a bunch of different collections. This behaviour allows someone to set a config variable which was a list, using a tuple, for instance, even though the default value was a list. I think its fine, but also being stricter would be fine too.

Reply:
I think I understand. If someone sets a value in config.py to a tuple or set, it will be coerced to a list (with a warning), so only lists are stored in and loaded from config.json, thus only allowing lists here.